src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1241,6 +1488,191 @@ private bool RequestUrlStartsWithPathBase(string requestUrl, out bool caseMatche
+                }
+
+                if (ch1 == ByteCR)
+                {
+                    // Check for final CRLF.
+                    if (ch2 == -1)
+                    {
+                        return false;
+                    }
+                    else if (ch2 == ByteLF)
+                    {
+                        consumed = reader.Cursor;
+                        examined = consumed;
+                        ConnectionControl.CancelTimeout();
+                        return true;
+                    }
+
+                    // Headers don't end in CRLF line.
+                    RejectRequest(RequestRejectionReason.HeadersCorruptedInvalidHeaderSequence);
+                }
+                else if (ch1 == ByteSpace || ch1 == ByteTab)
+                {
+                    RejectRequest(RequestRejectionReason.HeaderLineMustNotStartWithWhitespace);
+                }
+
+                // If we've parsed the max allowed numbers of headers and we're starting a new
+                // one, we've gone over the limit.
+                if (_requestHeadersParsed == ServerOptions.Limits.MaxRequestHeaderCount)
+                {
+                    RejectRequest(RequestRejectionReason.TooManyHeaders);
+                }
+
+                // Reset the reader since we're not at the end of headers
+                reader = start;
+
+                // Now parse a single header
+                ReadCursor lineEnd;
+                var limitedBuffer = buffer;
+                if (bufferLength >= _remainingRequestHeadersBytesAllowed)
+                {
+                    limitedBuffer = buffer.Slice(consumed, _remainingRequestHeadersBytesAllowed);
+                }
+
+                if (ReadCursorOperations.Seek(consumed, limitedBuffer.End, out lineEnd, ByteLF) == -1)
+                {
+                    if (limitedBuffer.Length == _remainingRequestHeadersBytesAllowed)
+                    {
+                        RejectRequest(RequestRejectionReason.HeadersExceedMaxTotalSize);
+                    }
+                    else
+                    {
+                        return false;
+                    }
+                }
+
+                const int stackAllocLimit = 512;
+
+                // Get a specific header line
+                lineEnd = limitedBuffer.Move(lineEnd, 2);
+                var headerBuffer = limitedBuffer.Slice(consumed, lineEnd);
+
+                Span<byte> span;
+                if (headerBuffer.IsSingleSpan)
+                {
+                    // No copies, directly use the one and only span
+                    span = headerBuffer.ToSpan();
+                }
+                else if (headerBuffer.Length < stackAllocLimit)
+                {
+                    unsafe
+                    {
+                        // Multiple buffers and < stackAllocLimit, copy into a stack buffer
+                        byte* stackBuffer = stackalloc byte[stackAllocLimit];
```

Comment:
No need to go full on?

`byte* stackBuffer = stackalloc byte[headerBuffer.Length];`?